ADR/ADR-003-PyBind11-For-CPP-Bindings.md

```diff
@@ -0,0 +1,295 @@
+
+Licence: GPLv3, generated code has same licencing as input to SWIG
+Maturity: settled and mature project, initial release 1996
+Longevity: 272 contributors (github)
+Activity: weekly commits, few patches per year
+Visibility: 6.1k stars on Github
+
+SWIG (Simplified Wrapper and Interface Generator) is a tool that automatically generates binding code to expose C/C++ functions and classes 
+to multiple languages, including Python. 
+
+An user-specified interface file (.i) describes what should be exposed, and SWIG generates the C/C++ glue code plus Python wrappers.
+
+Key Points:
+ * Supports many languages besides Python (Java, Ruby, etc.).
+ * Reduces boilerplate by generating the bindings for you.
+ * Handles C++ classes, functions, enums, and more.
+ * Requires SWIG tool installation and a simple build step.
+
+
+### Analysis
+
+We want to have python bindings for our whole C++ stack - this has a long term impact.
+From that derived requirements for a binding framework are: 
+ * longe term maintenance of the framework should be guaranteed for the foreseeable future
+ * low binding complexity to reduced development and maintenance overhead
+ * needs to be battle tested (due to the expected wide adoption)
+
+According to this we compare the canditates as follows:
+
+CPython:
+ * Pros: full control, no external dependencies, stable API, flexible and lightweight
+ * Cons: Boilerplate code, manual memory/error handling, steeper learning curve.
+
+CFFI:
+ * Pro: good for existing C interfaces
+ * Cons: needs additional C interface, bindings can get complex, higher maintenance cost
+
+Pybind11:
+ * Pros: low binding complexity, battle test in famous frameworks, interaction with numpy array
+ * Cons: higher compile-time due to code generation, larger binary size
+
+Nanobind:
+ * Pros: compensates pybind11 cons
+ * Cons: not header-only anymore, fewer `std` or third party support
+
+Boost.Python:
+ * Cons: clunky, heavy dependency on boost, larger build complexity, slower compile times
+
+Swig:
+ * Pros: automates most of the binding work, supports multiple languages, mature and stable tool
+ * Cons: interface files add an extra step, generated code may be harder to debug, 
+   less "pythonic" feel compared to pybind11 or Boost.Python
+
+
+Our canditate of choice is Pybind11 - compared to all the other candidates (except nanobind) it offers the 
+  * lowest development complexity
+  * gives direct access to complex objects like iterators
+  * is well maintained & documented
+  * battle-tested and widely used in famous frameworks
+
+Depending on the development on nanobind over the next coming years. 
+It might become a replacement for pybind11 - as the API is similar to pybind11, a migration will be rather simple.
+
+
+### Related Decisions
+
+None
+
+
+## Consequences
+
+Existing python bindings that are not pybind11 bindings can be migrated through the [design guidlines](#design-guidelines).
```

Comment:
Can _optionally_ be migrated. Or be migrated when convenient. There is not a significant pressure to do so.

Reply:
Done.